pkg/packetfilter/fake/packetfilter.go

```diff
@@ -0,0 +1,469 @@
+/*
+SPDX-License-Identifier: Apache-2.0
+
+Copyright Contributors to the Submariner project.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package fake
+
+import (
+	"strings"
+	"sync"
+
+	. "github.com/onsi/gomega"
+	"github.com/pkg/errors"
+	"github.com/submariner-io/submariner/pkg/ipset"
+	fakeipset "github.com/submariner-io/submariner/pkg/ipset/fake"
+	"github.com/submariner-io/submariner/pkg/packetfilter"
+	"github.com/submariner-io/submariner/pkg/packetfilter/iptables"
+	"k8s.io/utils/set"
+)
+
+const (
+	remoteCIDRIPSet                    = "SUBMARINER-REMOTECIDRS"
+	localCIDRIPSet                     = "SUBMARINER-LOCALCIDRS"
+	smPostRoutingChain                 = "SUBMARINER-POSTROUTING"
+	smGlobalnetEgressChainForPods      = "SM-GN-EGRESS-PODS"
+	smGlobalnetEgressChainForNamespace = "SM-GN-EGRESS-NS"
+	gnIPSetPrefix                      = "SM-GN-"
+	DriverName                         = "fake"
+)
+
+type globalEgressSet struct {
+	ipSetNamed ipset.Named
+	snatCIDR   string
+	markValue  string
+}
+
+type PacketFilter struct {
+	mutex                    sync.Mutex
+	chainRules               map[string]set.Set[string]
+	tableChains              map[string]set.Set[string]
+	failOnAppendRuleMatchers []interface{}
+	failOnDeleteRuleMatchers []interface{}
+
+	remoteIPSet             ipset.Named
+	localIPSet              ipset.Named
+	stringToGlobalEgressSet map[packetfilter.GlobalEgressHandle]*globalEgressSet
+	IPSetIface              ipset.Interface
+}
+
+var (
+	tableTypeToStr             = [packetfilter.TableTypeMAX]string{"filter", "mangle", "nat"}
+	iphookChainTypeToStr       = [packetfilter.ChainTypeMAX]string{"filter", "mangle", "nat"}
+	iphookChainTypeToTableType = [packetfilter.ChainTypeMAX]packetfilter.TableType{
+		packetfilter.TableTypeFilter,
+		packetfilter.TableTypeRoute,
+		packetfilter.TableTypeNAT,
+	}
+	chainHookToStr = [packetfilter.ChainHookMAX]string{"PREROUTING", "INPUT", "FORWARD", "OUTPUT", "POSTROUTING"}
+	driverInstance *PacketFilter
+)
+
+func ClearDriver() {
+	driverInstance = nil
+}
+
+func New() *PacketFilter {
+	if driverInstance != nil {
+		return driverInstance
+	}
```

Comment:
I don't think we should define a global instance. I did so initially in my PR but ran into issues due to the retained state (as I'm sure you did). You've added `ClearDriver` to alleviate that but that requires every consumer to explicitly call it. It's simpler and cleaner to have `New` return a new instance on each call. In my version I also call `SetNewDriverFn` for convenience.

```
func New() *PacketFilter {
	pf := &PacketFilter{
		chainRules:  map[string]set.Set[string]{},
		tableChains: map[string]set.Set[string]{},
	}

	packetfilter.SetNewDriverFn(func() (packetfilter.Driver, error) {
		return pf, nil
	})

	return pf
}
```